satellite/satellitedb/overlaycache.go

```diff
@@ -33,112 +33,170 @@ type overlaycache struct {
 
-	nodeType := int(pb.NodeType_STORAGE)
-
-	safeQuery := `
-		WHERE disqualified IS NULL
-		AND suspended IS NULL
-		AND exit_initiated_at IS NULL
-		AND type = ?
-		AND free_disk >= ?
-		AND total_audit_count >= ?
-		AND total_uptime_count >= ?
-		AND last_contact_success > ?`
-	args := append(make([]interface{}, 0, 13),
-		nodeType, criteria.FreeDisk, criteria.AuditCount,
-		criteria.UptimeCount, time.Now().Add(-criteria.OnlineWindow))
-
-	if criteria.MinimumVersion != "" {
-		v, err := version.NewSemVer(criteria.MinimumVersion)
+	if newNodeCount == 0 && reputableNodeCount == 0 {
+		return nil, nil
+	}
+	safeNewNodeQuery := ""
+	newNodeArgs := []interface{}{}
+	if newNodeCount > 0 {
+		safeNewNodeQuery, newNodeArgs, err = buildConditions(ctx, criteria, true)
 		if err != nil {
-			return nil, Error.New("invalid node selection criteria version: %v", err)
+			return nil, err
 		}
-		safeQuery += `
-			AND (major > ? OR (major = ? AND (minor > ? OR (minor = ? AND patch >= ?))))
-			AND release`
-		args = append(args, v.Major, v.Major, v.Minor, v.Minor, v.Patch)
+	}
+	safeReputableNodeQuery, reputableNodeArgs, err := buildConditions(ctx, criteria, false)
+	if err != nil {
+		return nil, err
 	}
 
 	if !criteria.DistinctIP {
-		nodes, err = cache.queryNodes(ctx, criteria.ExcludedIDs, count, safeQuery, args...)
+		newNodeQuery, moreNewNodeArgs := buildSelection(ctx, newNodeCount, safeNewNodeQuery)
+		newNodeArgs = append(newNodeArgs, moreNewNodeArgs...)
+
+		reputableNodeQuery, moreReputableNodeArgs := buildSelection(ctx, reputableNodeCount, safeReputableNodeQuery)
+		reputableNodeArgs = append(reputableNodeArgs, moreReputableNodeArgs...)
+
+		finalQuery := newNodeQuery + " UNION ALL " + reputableNodeQuery
+		finalArgs := []interface{}{}
+		finalArgs = append(finalArgs, newNodeArgs...)
+		finalArgs = append(finalArgs, reputableNodeArgs...)
+
+		rows, err := cache.db.Query(ctx, cache.db.Rebind(finalQuery), finalArgs...)
 		if err != nil {
-			return nil, err
+			return nil, Error.Wrap(err)
+		}
+		defer func() { err = errs.Combine(err, rows.Close()) }()
+
+		var nodes []*overlay.SelectedNode
+		for rows.Next() {
+			var node overlay.SelectedNode
+			node.Address = &pb.NodeAddress{Transport: pb.NodeTransport_TCP_TLS_GRPC}
+			err = rows.Scan(&node.LastNet, &node.ID, &node.Address.Address, &node.LastIPPort)
+			if err != nil {
+				return nil, err
+			}
+			nodes = append(nodes, &node)
 		}
 		return nodes, nil
 	}
 
+	totalCount := newNodeCount + reputableNodeCount
+	receivedNewNodeCount := 0
+	receivedNodeNetworks := make(map[string]struct{})
+
 	for i := 0; i < 3; i++ {
-		moreNodes, err := cache.queryNodesDistinct(ctx, criteria.ExcludedIDs, criteria.ExcludedNetworks, count-len(nodes), safeQuery, criteria.DistinctIP, args...)
+		finalQuery := ""
+		finalArgs := []interface{}{}
+
+		// temporary arg slices are needed for each iteration, otherwise we'll end up with too many args appended
+		tempNewNodeArgs := []interface{}{}
```

Comment:
This is mixing query building and query execution layers making this hard to follow.